### PR TITLE
[docs] Add initial AIR documentation

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Going from 'bare bones' documentation to something that is more of a usable set of documents.

This is still largely setting up the structure; we will add much more info as we go forward.

https://docs.google.com/document/d/1mr-vT4xNqlU-SqLDUUo8waA9pIZFxms438QJTKy5dqc/edit#heading=h.muqmahorokn3

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/_toc.yml [hunk 1]
```diff
@@ -105,6 +105,21 @@ parts:
           - file: workflows/package-ref
       - file: ray-more-libs/index
         title: More Ray ML Libraries
```
Comment (maxpumperla) [inline doc/source/_toc.yml:107]: Can we create an `images` folder inside of `ray-air`? I think that's cleaner (we have too many images in there already, not clear what belongs where)

File: doc/source/_toc.yml [hunk 2]
```diff
@@ -105,6 +105,21 @@ parts:
...
+        sections:
+          - file: ray-air/getting-started
+            sections:
+              - file: ray-air/key-concepts
+              - file: ray-air/deployment
+              - file: ray-air/package-ref
+          - file: ray-more-libs/joblib
+          - file: ray-more-libs/lightgbm-ray
+          - file: ray-more-libs/multiprocessing
```
Comment (maxpumperla) [inline doc/source/_toc.yml:116]: +1 for making the TOC explicit

File: doc/source/ray-air/deployment.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+-----------------
+
+* Ray AIR handles the heavyweight compute aspects of AI apps and services.
+* Ray AIR relies on external integrations (e.g., Tecton, MLFlow, W&B) for Storage and Tracking.
+* Workflow Orchestrators (e.g., AirFlow) are an optional component that can be used for scheduling recurring jobs, launching new Ray clusters for jobs, and running non-Ray compute steps.
+* Lightweight orchestration of task graphs within a single Ray AIR app can be handled with Ray Workflows.
+* Ray AIR libraries can be used independently, within an existing ML platform, or to build a Ray-native ML platform.
+
+
+Ray AIR libraries à la carte
```
Comment (maxpumperla) [inline doc/source/ray-air/deployment.rst:20]: let's be clear first, witty later. not sure many users will understand this term. maybe "Pick the libraries you need"

File: doc/source/ray-air/deployment.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+This is applicable to you if you are an ML engineer wants to independently use a Ray AIR library for a specific AI app or service use case and do not need to integrate with an existing ML platform.
+
+For example, Alice wants to use RLlib to train models for her work project. Bob wants to use Ray Serve to deploy his model pipeline. In both cases, Alice and Bob can leverage these libraries independently without any coordination.
+
+This scenario describes most usages of Ray libraries today.
+
+.. image:: /images/air_arch_1.png
+
+
+Highlighted points
```
Comment (maxpumperla) [inline doc/source/ray-air/deployment.rst:34]: Bullet points are highlighted points, so that's kind of understood. Can we maybe just lead over to the "highlights" with a quick sentence instead?

Also, I don't feel like these two points will help users much in this context. We don't really know what Autoscaler, Anyscale or KubeRay are at this point (this is all high-level on purpose up to this point). How does pick-and-choose of libraries relate to managed clusters?

I'd also suggest to remove "GCS", unless we want to explain what it is or link to an explanation.

File: doc/source/ray-air/deployment.rst [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+.. image:: /images/air_arch_2.png
+
+
+Highlighted points
+~~~~~~~~~~~~~~~~~~
+
+1. A workflow orchestrator such as AirFlow, Oozie, SageMaker Pipelines, etc. is responsible for scheduling and creating Ray AIR clusters and running AIR apps and services. The AIR app may be part of a larger orchestrated workflow (e.g., Spark ETL, then Training on Ray).
+2. Lightweight orchestration of task graphs can be handled entirely within AIR using Ray Workflows. The workflow orchestrator is only needed if running non-Ray steps.
```
Comment (maxpumperla) [inline doc/source/ray-air/deployment.rst:57]: Link to workflows?

**Comment by maxpumperla** [comment]
Screenshot 2022-05-05 at 09 02 35

the more libs section does not render as we want now

File: doc/source/ray-air/key-concepts.rst [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        batch_predictor.predict(test_dataset)
+        .map_batches(lambda df: (df > 0.5).astype(int), batch_format="pandas")
+        .to_pandas(limit=float("inf"))
+    )
+
+
+Online Inference
+----------------
+
+TODO
```
Comment (maxpumperla) [inline doc/source/ray-air/key-concepts.rst:129]: should we hide this one as long as it's a TODO?

File: doc/source/ray-air/key-concepts.rst [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+Preprocessors
+-------------
+
+Preprocessors are primitives that can be used to transform input data into features.
+
+A preprocessor can be fitted during Training, and applied at runtime in both Training and Serving on data batches in the same way.
+
+Preprocessors operate on :ref:`Ray Datasets <datasets>`.
+
+.. code-block:: python
```
Comment (maxpumperla) [inline doc/source/ray-air/key-concepts.rst:19]: can we please add these snippets into scripts in a `doc_code` folder and test that in CI? We don't want code directly in docs anymore (unless using notebooks). Let's start this off the right way! :D

File: doc/source/ray-air/getting-started.rst [hunk 8]
```diff
@@ -3,147 +3,71 @@ Ray AI Runtime (alpha)
...
-Ray AIR consists of 5 key components -- Data processing (Ray Data), Model Training (Ray Train), Reinforcement Learning (Ray RLlib), Hyperparameter Tuning (Ray Tune), and Model Serving (Ray Serve).
+Ray AIR consists of 5 key components -- Data processing (:ref:`Ray Data <datasets>`), Model Training (:ref:`Ray Train <train-docs>`), Reinforcement Learning (:ref:`Ray RLlib <rllib-index>`), Hyperparameter Tuning (:ref:`Ray Tune <tune-main>`), and Model Serving (:ref:`Ray Serve <rayserve>`).
 
 Users can use these libraries interchangeably to scale different parts of standard ML workflows.
 
 
 .. tip::
     **Getting involved with Ray AIR.** We'll be holding office hours, development sprints, and other activities as we get closer to the Ray AIR Beta/GA release. Want to join us? Fill out `this short form <https://forms.gle/wCCdbaQDtgErYycT6>`__!
 
+Quick Start
```
Comment (maxpumperla) [inline doc/source/ray-air/getting-started.rst:23]: Love the quick start ❤️

File: doc/source/ray-more-libs/index.rst [hunk 9]
```diff
@@ -5,17 +5,6 @@ More Ray ML Libraries
...
-    multiprocessing
-    ray-collective
-    ray-lightning
-    ../ray-core/examples/using-ray-with-pytorch-lightning
-    xgboost-ray
-    ../ray-core/examples/dask_xgboost/dask_xgboost
-    ../ray-core/examples/modin_xgboost/modin_xgboost
\ No newline at end of file
+.. contents::
+    :depth: 1
```
Comment (maxpumperla) [inline doc/source/ray-more-libs/index.rst:10]: this seems to be the wrong depth, see screenshot in main comment section

File: doc/source/ray-air/package-ref.rst [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+Serving
+~~~~~~~
+
+.. autoclass:: ray.serve.model_wrappers.ModelWrapperDeployment
+
+.. autoclass:: ray.serve.model_wrappers.ModelWrapper
+
+.. _air-results-ref:
+
+Outputs
```
Comment (maxpumperla) [inline doc/source/ray-air/package-ref.rst:120]: Probably not for this PR (too low prio), but the anchors for Outputs and Configs don't seem to work as expected. The nav "hangs" at Serving when scrolling down.

Screenshot 2022-05-05 at 09 12 35

**Review by maxpumperla — CHANGES_REQUESTED** [review]
super cool! just a couple of nits, @richardliaw

File: doc/source/ray-air/key-concepts.rst [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+.. literalinclude:: doc_code/key_concepts.py
+    :language: python
+    :start-after: __air_batch_predictor_start__
+    :end-before: __air_batch_predictor_end__
+
+
+.. Online Inference
+.. ----------------
+
+.. TODO
```
Comment (richardliaw) [inline doc/source/ray-air/key-concepts.rst:76]: @simon-mo is this checked in yet? do you have an example for using modelwrappers?

Comment (simon-mo) [reply doc/source/ray-air/key-concepts.rst:76]: not yet. can you create a ticket and assign to me? so far @krfricke has an end to end example https://github.com/ray-project/ray/commit/76da2255d9c59d485e8cd5e942de49eb93e03011

File: doc/source/ray-air/doc_code/key_concepts.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+from ray.tune import Tuner, TuneConfig
+
+tuner = Tuner(
+    trainer,
+    param_space={
+        "params": {
+            "max_depth": tune.randint(1, 9)
+        }
+    },
+    tune_config=TuneConfig(num_samples=20, metric="loss", mode="min"),
```
Comment (krfricke) [inline doc/source/ray-air/doc_code/key_concepts.py:63]: If we're still in the context of the XGBoost trainer, the loss has a slightly different name

```suggestion
    tune_config=TuneConfig(num_samples=20, metric="train-logloss", mode="min"),
```

File: doc/source/ray-air/doc_code/key_concepts.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+batch_predictor = BatchPredictor.from_checkpoint(
+    result.checkpoint, XGBoostPredictor
+)
+
+predicted_labels = (
+    batch_predictor.predict(test_dataset)
+    .map_batches(lambda df: (df > 0.5).astype(int), batch_format="pandas")
+    .to_pandas(limit=float("inf"))
+)
```
Comment (krfricke) [inline doc/source/ray-air/doc_code/key_concepts.py:80]: Out of scope of this PR, but I think it's not ideal to have these transformations after prediction. Maybe we can add this functionality as a flag to the `XGBoostPredictor` so we can get rid of this clutter. cc @Yard1

File: doc/source/ray-air/getting-started.rst (hunk 8 above)
Comment (krfricke) [reply doc/source/ray-air/getting-started.rst:23]: I love that we have a quickstart but I don't love the examples, they don't seem very "quick". But that's our of scope for this PR, let's check in the documentation structure first and iterate on the examples later

File: doc/source/ray-air/getting-started.rst [hunk 14]
```diff
@@ -3,147 +3,71 @@ Ray AI Runtime (alpha)
...
-    :members:
-    :show-inheritance:
+.. dropdown:: Multi-cloud and Framework-interoperable
+    :animate: fade-in-slide-down
 
-.. automodule:: ray.ml.predictors.integrations.lightgbm
-    :members:
-    :show-inheritance:
+    Ray AIR is multi-cloud and framework-interoperable. The Ray compute layer and libraries freely operate with Cloud platforms and frameworks in the ecosystem, reducing lock-in to any particular choices of ML tech.
+    Ex: why framework interoperability is unique to Ray--- it's easy to run Torch distributed or elastic Horovod within Ray, but not vice versa.
```
Comment (krfricke) [inline doc/source/ray-air/getting-started.rst:59]: This seems to be a placeholder

**Review by krfricke — COMMENTED** [review]
This is a great start for the documentation.

One thing we could maybe add is a section that clearly explains "What is the difference between Ray AIR and the other Ray libraries" - I believe for existing users it's not immediately clear that AIR acts as an interface for existing functionality.

File: doc/source/ray-air/deployment.rst (hunk 5 above)
Comment (richardliaw) [reply doc/source/ray-air/deployment.rst]: Updated.

File: doc/source/ray-air/deployment.rst (hunk 4 above)
Comment (richardliaw) [reply doc/source/ray-air/deployment.rst]: Updated!

File: doc/source/ray-air/key-concepts.rst (hunk 6 above)
Comment (richardliaw) [reply doc/source/ray-air/key-concepts.rst]: done

**Comment by ericl** [comment]
@richardliaw for the quick start, how about breaking them down into smaller snippets with intermediate explanatory text? I believe tabs can hold multi-part docs.

Example:

Prepare data...
[code]

Do training...
[code]

Generate predictions...
[code]

File: doc/source/ray-air/doc_code/key_concepts.py (hunk 13 above)
Comment (Yard1) [reply doc/source/ray-air/doc_code/key_concepts.py:80]: The output of XGboost's predict method depends on the objective. I don't think we can cover all cases. Leaving it up to the user seems fine to me. Maybe we can show how to extend the predictor class to achieve it. It would be more performant, too.

File: doc/source/ray-air/deployment.rst [hunk 15]
```diff
@@ -0,0 +1,58 @@
+.. _air-deployment:
+
+AIR Deployment Guide
```
[the remark this answers is not in the document]
Comment (maxpumperla) [reply doc/source/ray-air/deployment.rst:3]: sure, whatever works. consistency is important, whichever way we want to go

**Review by maxpumperla — APPROVED** [review]

**Comment by richardliaw** [comment]
TODO for myself: 

1. break up quickstart examples
2. address kai nits
3. make sure code is working.

File: doc/BUILD [hunk 16]
```diff
@@ -159,6 +159,20 @@ py_test_run_all_subdirectory(
     tags = ["exclusive", "team:ml"],
 )
 
+
+# --------------------------------------------------------------------
+# Test all doc/source/ray-air/doc_code code included in rst/md files.
+# --------------------------------------------------------------------
+
+py_test_run_all_subdirectory(
+    size = "medium",
```
Comment (richardliaw) [inline doc/BUILD:168]:
```suggestion
    size = "large",
```
